### PR TITLE
MGMT-7331: Validate host platform

### PR DESCRIPTION
**Pull request by eliorerz**
Validate host platform by comparing host supported platforms and cluster platform property.

## List all the issues related to this PR

- [x] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [x] Cloud
- [ ] Operator Managed Deployments
- [x] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [ ] Waiting for CI to do a full test run
- [x] Manual (Elaborate on how it was tested)
- [ ] No tests needed

/cc @gamli75 
/cc @filanov 
/cc @michaellevy101 

---
validation_message
Insufficient_hosts

### CODE REVIEW

File: internal/host/validator.go [hunk 1]
```diff
@@ -243,6 +243,35 @@ func (v *validator) hasMinMemory(c *validationContext) ValidationStatus {
...
+	return boolValue(false)
+}
+
+func (v *validator) printHasSupportedPlatform(c *validationContext, status ValidationStatus) string {
+	switch status {
+	case ValidationSuccess:
+		return fmt.Sprintf("Platform %s is supported by the cluster", c.cluster.Platform.Type)
+	case ValidationFailure:
+		hostAvailablePlatforms := common.GetSupportedPlatforms(*c.inventory)
+		return fmt.Sprintf("The host is not eligible to participate in Openshift Cluster because the host platform (%v) is not competible by the cluster platform (%s)",
```
Comment (romfreiman) [inline internal/host/validator.go:266]: compatible
OpenShift

File: internal/host/hostutil/update_host_test.go [hunk 2]
```diff
@@ -43,7 +43,7 @@ var _ = Describe("update_host_state", func() {
 		lastUpdatedTime = host.StatusUpdatedAt
 	})
 
-	Describe("UpdateHostStatus", func() {
+	FDescribe("UpdateHostStatus", func() {
```
Comment (michaellevy101) [inline internal/host/hostutil/update_host_test.go]: Describe

File: internal/common/platform.go [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+func GetSupportedPlatforms(inventory models.Inventory) *[]models.PlatformType {
+	hostPlatform := []models.PlatformType{models.PlatformTypeBaremetal}
+	AppendPlatformFromVendor(*inventory.SystemVendor, &hostPlatform)
+	return &hostPlatform
+}
+
+func AppendPlatformFromVendor(vendor models.SystemVendor, hostsPlatform *[]models.PlatformType) {
+	platform := supportedManufacturers[vendor.Manufacturer]
+	if platform != "" {
+		*hostsPlatform = append(*hostsPlatform, platform)
```
Comment (michaellevy101) [inline internal/common/platform.go:22]: GetHostPlatform - try to write a immutable functions   
append in the GetSupportedPlatforms function

File: internal/host/hostutil/update_host_test.go [hunk 4]
```diff
@@ -82,7 +82,7 @@ var _ = Describe("update_host_state", func() {
 			})
 		})
 
-		It("db_failure", func() {
+		FIt("db_failure", func() {
```
Comment (michaellevy101) [inline internal/host/hostutil/update_host_test.go]: It

File: internal/host/validator.go [hunk 5]
```diff
@@ -243,6 +243,35 @@ func (v *validator) hasMinMemory(c *validationContext) ValidationStatus {
...
 
+func (v *validator) hasSupportedPlatform(c *validationContext) ValidationStatus {
+	if c.inventory == nil || c.cluster.Platform.Type == "" {
+		return ValidationPending
+	}
+	hostAvailablePlatforms := common.GetSupportedPlatforms(*c.inventory)
+
+	for _, platform := range *hostAvailablePlatforms {
+		if c.cluster.Platform.Type == platform {
+			return boolValue(true)
```
Comment (michaellevy101) [inline internal/host/validator.go:254]: boolValue?

File: internal/host/validator.go (hunk 1 above)
Comment (michaellevy101) [inline internal/host/validator.go:266]: not sure about the message, it's too wordy, @avishayt  what do you think?

File: internal/host/validator.go [hunk 6]
```diff
@@ -243,6 +243,35 @@ func (v *validator) hasMinMemory(c *validationContext) ValidationStatus {
...
+		}
+	}
+	return boolValue(false)
+}
+
+func (v *validator) printHasSupportedPlatform(c *validationContext, status ValidationStatus) string {
```
Comment (michaellevy101) [inline internal/host/validator.go:260]: maybe the function name/validation name should be compatibleWithClusterPlatform

**Review by michaellevy101 — COMMENTED** [review]
Looks good!!! 
Please add a detailed description in the commit and the PR description
little bit of background and reasoning behind the decisions
The Jira ticket isn't available from the IDE so to help the developer to understand the reasoning behind the code we should describe  the solution in the commit message which available to him

File: internal/host/validator.go (hunk 6 above)
Comment (eliorerz) [reply internal/host/validator.go:260]: This is the convention on this file, print + <validationID>

File: internal/host/validator.go (hunk 5 above)
Comment (eliorerz) [reply internal/host/validator.go:254]: This is the convention on this file, it returns `ValidationSuccess`

**Review by filanov — COMMENTED** [review]
please add failure and pending tests

File: swagger.yaml [hunk 7]
```diff
@@ -5058,6 +5058,7 @@ definitions:
       - 'sufficient-network-latency-requirement-for-role'
       - 'sufficient-packet-loss-requirement-for-role'
       - 'has-default-route'
+      - 'has-supported-platform'
```
Comment (avishayt) [inline swagger.yaml:5061]:
```suggestion
      - 'compatible-with-cluster-platform'
```

File: internal/host/validator.go [hunk 8]
```diff
@@ -243,6 +243,35 @@ func (v *validator) hasMinMemory(c *validationContext) ValidationStatus {
...
+			return ValidationSuccess
+		}
+	}
+	return ValidationFailure
+}
+
+func (v *validator) printHasSupportedPlatform(c *validationContext, status ValidationStatus) string {
+	switch status {
+	case ValidationSuccess:
+		return fmt.Sprintf("Platform %s is supported by the cluster", c.cluster.Platform.Type)
```
Comment (avishayt) [inline internal/host/validator.go:263]:
```suggestion
		return fmt.Sprintf("Host is compatible with cluster platform %s", c.cluster.Platform.Type)
```

File: internal/host/validator.go [hunk 9]
```diff
@@ -243,6 +243,35 @@ func (v *validator) hasMinMemory(c *validationContext) ValidationStatus {
...
+	case ValidationFailure:
+		hostAvailablePlatforms := common.GetSupportedPlatforms(*c.inventory)
+		return fmt.Sprintf("The host is not eligible to participate in Openshift Cluster because the host platform (%v) is not competible by the cluster platform (%s)",
+			*hostAvailablePlatforms, c.cluster.Platform.Type)
```
Comment (avishayt) [inline internal/host/validator.go:267]:
```suggestion
		return fmt.Sprintf("Host is not compatible with cluster platform %s; either disable this host or choose a compatible cluster platform (%v)",
			c.cluster.Platform.Type, *hostAvailablePlatforms)
```

File: internal/host/statemachine.go [hunk 10]
```diff
@@ -502,7 +502,7 @@ func NewHostStateMachine(th *transitionHandler) stateswitch.StateMachine {
 		PostTransition:   th.PostRefreshHost(statusInfoDiscovering),
 	})
 
-	var hasMinRequiredHardware = stateswitch.And(If(HasMinValidDisks), If(HasMinCPUCores), If(HasMinMemory), If(IsPlatformValid))
+	var hasMinRequiredHardware = stateswitch.And(If(HasMinValidDisks), If(HasMinCPUCores), If(HasMinMemory), If(IsPlatformValid), If(CompatibleWithClusterPlatform))
```
Comment (filanov) [inline internal/host/statemachine.go:505]: please break the line.

**Review by filanov — APPROVED** [review]

**Comment by filanov** [comment]
/hold 
@eliorerz does it depend on MGMT-7280 (it's reverted at the moment)

**Comment by avishayt** [comment]
I don't see any new unit tests added

**Comment by eliorerz** [comment]
> /hold
> @eliorerz does it depend on [MGMT-7280](https://issues.redhat.com/browse/MGMT-7280) (it's reverted at the moment)

@filanov 
/unhold
This PR is standalone but there are 2 duplicate files - `common.platform.go` and `common.inventory.go` .
Rebase will be needed after that https://github.com/openshift/assisted-service/pull/2306 is merged

**Review by avishayt — APPROVED** [review]

**Review by gamli75 — APPROVED** [review]
